src/bip39.cpp

```diff
@@ -0,0 +1,179 @@
+        }
+        strcpy(p, wordlist[idx]);
+        p += strlen(wordlist[idx]);
+        *p = (i < mlen - 1) ? ' ' : 0;
+        p++;
+    }
+
+    return mnemo;
+}
+
+int mnemonic_check(const char *mnemonic)
+{
+    if (!mnemonic) {
+        return 0;
+    }
+
+    uint32_t i, n;
+
+    i = 0; n = 0;
+    while (mnemonic[i]) {
+        if (mnemonic[i] == ' ') {
+            n++;
+        }
+        i++;
+    }
+    n++;
+    // check number of words
+    if (n != 12 && n != 18 && n != 24) {
+        return 0;
+    }
+
+    char current_word[10];
+    uint32_t j, k, ki, bi;
+    uint8_t bits[32 + 1];
+    memset(bits, 0, sizeof(bits));
+    i = 0; bi = 0;
+    while (mnemonic[i]) {
+        j = 0;
+        while (mnemonic[i] != ' ' && mnemonic[i] != 0) {
+            if (j >= sizeof(current_word) - 1) {
+                return 0;
+            }
+            current_word[j] = mnemonic[i];
+            i++; j++;
+        }
+        current_word[j] = 0;
+        if (mnemonic[i] != 0) i++;
+        k = 0;
+        for (;;) {
+            if (!wordlist[k]) { // word not found
+                return 0;
+            }
+            if (strcmp(current_word, wordlist[k]) == 0) { // word found on index k
+                for (ki = 0; ki < 11; ki++) {
+                    if (k & (1 << (10 - ki))) {
+                        bits[bi / 8] |= 1 << (7 - (bi % 8));
+                    }
+                    bi++;
+                }
+                break;
+            }
+            k++;
+        }
+    }
+    if (bi != n * 11) {
+        return 0;
+    }
+    bits[32] = bits[n * 4 / 3];
+    // sha256_Raw(bits, n * 4 / 3, bits);
+    CSHA256().Write(bits, n * 4 / 3).Finalize(bits);
+
+    if (n == 12) {
+        return (bits[0] & 0xF0) == (bits[32] & 0xF0); // compare first 4 bits
+    } else
+    if (n == 18) {
+        return (bits[0] & 0xFC) == (bits[32] & 0xFC); // compare first 6 bits
+    } else
+    if (n == 24) {
+        return bits[0] == bits[32]; // compare 8 bits
+    }
+    return 0;
+}
+
+// passphrase must be at most 256 characters or code may crash
+void mnemonic_to_seed(const char *mnemonic, const char *passphrase, uint8_t seed[512 / 8], void (*progress_callback)(uint32_t current, uint32_t total))
+{
+    int passphraselen = strlen(passphrase);
+    uint8_t salt[8 + 256 + 4];
+    memcpy(salt, "mnemonic", 8);
+    memcpy(salt + 8, passphrase, passphraselen);
+    // pbkdf2_hmac_sha512((const uint8_t *)mnemonic, strlen(mnemonic), salt, passphraselen + 8, BIP39_PBKDF2_ROUNDS, seed, 512 / 8, progress_callback);
+    // int PKCS5_PBKDF2_HMAC(const char *pass, int passlen,
+    //                    const unsigned char *salt, int saltlen, int iter,
+    //                    const EVP_MD *digest,
+    //                    int keylen, unsigned char *out);
+    PKCS5_PBKDF2_HMAC(mnemonic, strlen(mnemonic), salt, passphraselen + 8, 2048, EVP_sha512(), 64, seed);
+}
+
+const char * const *mnemonic_wordlist(void)
```

Comment:
This function isn't used.
It 'leaks' the representation as an array so would prevent using other representation such as a trie.
(Also, what's with the `void` parameter in a cpp file.)
I suggest we remove it, from both cpp and h..